src/App.tsx

```diff
@@ -1,26 +1,390 @@
+  const [deletingTodoId, setDeletingTodoId] = useState<number | null>(null);
+  const [updatingTodoId, setUpdatingTodoId] = useState<number | null>(null);
+
+  const loadTodos = async () => {
+    setIsLoading(true);
+    setErrorMessage('');
+    setShowError(false);
+    try {
+      const loadedTodos = await getTodos();
+
+      setTodos(loadedTodos);
+    } catch (error) {
+      setErrorMessage('Unable to load todos');
+      setShowError(true);
+      setTimeout(() => setShowError(false), 3000);
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    loadTodos();
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+
+    return () => {
+      setErrorMessage('');
+    };
+  }, []);
+
+  useEffect(() => {
+    if (!isAddingTodo && inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [isAddingTodo]);
+
+  const handleCreateTodo = async (e: React.FormEvent) => {
+    e.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      setErrorMessage('Title should not be empty');
+      setShowError(true);
+      setTimeout(() => setShowError(false), 3000);
+
+      return;
+    }
+
+    setIsAddingTodo(true);
+    setErrorMessage('');
+    setShowError(false);
+
+    const temp: Todo = {
+      id: 0,
+      userId: USER_ID,
+      title: newTodoTitle.trim(),
+      completed: false,
+    };
+
+    setTempTodo(temp);
+
+    try {
+      const newTodo = await createTodo(newTodoTitle);
+
+      setTodos(prevTodos => [...prevTodos, newTodo]);
+      setNewTodoTitle('');
+    } catch (error) {
+      setErrorMessage('Unable to add a todo');
+      setShowError(true);
+      setTimeout(() => {
+        setShowError(false);
+        setErrorMessage('');
+      }, 3000);
+    } finally {
+      setIsAddingTodo(false);
+      setTempTodo(null);
+      if (inputRef.current) {
+        inputRef.current.focus();
+      }
+    }
+  };
+
+  const handleDeleteTodo = async (id: number) => {
+    setDeletingTodoId(id);
+    setErrorMessage('');
+    setShowError(false);
+    try {
+      await deleteTodo(id);
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+    } catch (error) {
+      setErrorMessage('Unable to delete a todo');
+      setShowError(true);
+      setTimeout(() => setShowError(false), 3000);
+    } finally {
+      setDeletingTodoId(null);
+      if (inputRef.current) {
+        inputRef.current.focus();
+      }
+    }
+  };
+
+  const handleClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    setErrorMessage('');
+    setShowError(false);
+
+    const deletePromises = completedTodos.map(todo =>
+      deleteTodo(todo.id)
+        .then(() => {
+          return todo.id;
+        })
+        .catch(() => {
+          setErrorMessage('Unable to delete a todo');
+          setShowError(true);
+
+          return null;
+        }),
+    );
+
+    const results: (number | null)[] = await Promise.all(deletePromises);
```

Comment:
To pass all tests you should handle each result separately. Promise.allSettled will be better for it